Tag Page, Group Report Tag Add, Individual Report Tag Add

limmabean opened the pull request:
The big boy is here. But not without a bunch of slight hacking. I followed the current infrastructure standards as best as I could. There are 3 TODOs located throughout the codebase, none of which cause damage to functionality, but more for better code formatting. Please make efforts to call out and fix code structure, as I have been writing this at light speed.

Good luck, and comment a lot.

Commit:
Revert "Revert "test: SMTC model""

This reverts commit 5a338b3259f5b119ab08e31d63832a21a8e04cbf.

Commit:
Revert "Revert "test: API endpoints for SMTC tags""

This reverts commit 56636b92af0db891f1a03396ee0b6b9c57e017c2.

Commit:
Revert "Revert "feat: expose API endpoints for SMTC tags""

This reverts commit be9a6606b400f9e302936934e7cae6f147ee975e.

Commit:
Revert "Revert "feat: add SMTCTag model for report-level tagging""

This reverts commit 2145c3b05be4117d8b1c1f9f8d63a0812ace9525.

Commit:
Revert "Revert "test: add SMTC model to general db setup & teardown routines""

This reverts commit e752a156a4c57e9325b8b325dc4faa776fd31907.

Commit:
Bringing back Tags

Also adding Reports/Last min instead of Reports/min

Commit:
Merge branch 'ML-tag-frontend' into ML-Tags-Actual

mkyaw6 committed:
Add service to fetch, add and delete SMTC tags

Commit:
Added tagify stylesheets and removed select2 libraries

We don't actually use select2 for anything so I have removed the library.

Commit:
Merge branch 'ML-Tags-Actual' of https://github.com/TID-Lab/aggie into ML-Tags-Actual

mkyaw6 committed:
Add frontend updates for SMTC Tag operations

mkyaw6 committed:
Fix typo

Commit:
Remove Socket Listeners so we don't create leaks

Commit:
Merge branch 'ML-Tags-Actual' of https://github.com/TID-Lab/aggie into ML-Tags-Actual

Commit:
Merge branch 'ML-tag-frontend' into ML-Tags-Actual

Commit:
Merge branch 'ML-tag-frontend' into ML-Tags-Actual

mkyaw6 committed:
Use singular SMTCTag for consistency with backend

Commit:
Added Initial Route to Tags Page

Commit:
Merge branch 'ML-Tags-Actual' of https://github.com/TID-Lab/aggie into ML-Tags-Actual

karpawich committed:
fix: correct schema reference typo

karpawich committed:
Merge branch '333/report-level-tags' into ML-Tags-Actual

Commit:
Adding Tag Page, Modal, Controllers

Commit:
Removed Old Test SMTC Language

Commit:
Merge branch 'ML-tag-frontend' into ML-Tags-Actual

Commit:
Added Tag Pages and Modals

mkyaw6 committed:
Fix fronted tag display to use socket.io

Commit:
Move to indiana

Commit:
Merge branch 'develop' into ML-Tags-Actual

# Conflicts:
#	public/angular/templates/side-statsbar.html

Commit:
Got Descriptioin and Color Inputs Working

TODO: Get Creator and Date Created Working.

cooperka committed:
Lighten the gray color; use for background too

cooperka committed:
Also remove gray for read reports

Commit:
Added username to tag table and Updated Alert Messages

Also removed date and fixed response codes.

Commit:
Merge remote-tracking branch 'origin/ML-tags-Actual' into ML-tags-Actual

Commit:
Added Update Tags Feature

However the changes made to tags are not reflected until refresh. This will be fixed in a refactor of code to match the way incidents are updated and added. We must also consider the case when tags are deleted. They must also be deleted on every report they are on.

Commit:
Added coloring to Report Page Tags and Edited Tags Modal to remove descriptions

TODO: I need to add colors to Tags Modal

Commit:
Changed Colors and made grey colors schemes match Facebook

Commit:
Merge branch 'develop' into ML-tags-Actual

# Conflicts:
#	public/angular/js/controllers/analysis.js
#	public/angular/js/controllers/incidents/index.js
#	public/angular/js/controllers/navbar.js

Commit:
Merge branch 'develop' into ML-tags-Actual

Commit:
Merge branch 'develop' into ML-tags-Actual

Commit:
Removed Tabs Column from Tag Modal

This is redundant as the tags will be shown in the add & remove tags bar.

Commit:
Added Report API endpoint for SMTCTags

This updates the API to be able to add multiple Tags with one API request and this also matches the current method of adding and removing Incidents.

Commit:
Re ordering the Alert Bar

Commit:
Clear Tags from Selected Functionality

Added ability to clear tags from selected.

Commit:
Merge branch 'develop' into ML-tags-Actual

# Conflicts:
#	public/angular/templates/incidents/index.html

Commit:
Removing Secondary Color from Search Bars

This coloring isn't actually useful and would be better utilized in tags.

Commit:
Merge branch 'develop' into ML-tags-Actual

Commit:
Merge branch 'develop' into ML-tags-Actual

Commit:
Finalized Reports

Added measures to prevent duplicating tags and removing tags that aren't there. Also realized that tag.toString was necessary because of the object type in the schema file.

Commit:
Added Color Tags to Report Table and Adjusted Models

Changed the modals so that you can see information when adding incidents or tags. Have not added smtcTag functionally for individual reports on the Index page or the show page.

Commit:
Merge branch 'develop' into ML-tags-Actual

Commit:
Tags introduced to Add Tag to Report Modal

The functionality to actually change tags is still TODO. The next commit with include this.

Commit:
Individual Tag Editing for Reports

There are remaining TODOs but tags now can be added by report (similiar to incidents).

Commit:
Cursor: pointer the Edit and Remove Icons on Tag and Incident Page

The cursor didn't turn into a pointer when hovering over edit and delete icon.

cooperka reviewed (commented):
We can deploy this after rough code review so that folks can use it right away; before we merge to develop though, I do think it needs more tests.

Nice work overall, this has been a long time coming and I'm so excited!

List of random things I found, nothing specifically blocking release:
- Tags page > Add tag > Submit: button doesn't work if you haven't chosen a color (error message would be helpful, and/or default to a random color to make it one less step)
- Edit tag > Save:
    - Old info doesn't update until you manually refresh the page
    - Alert message says `smtcTag.update.success` instead of real text
- Reports list > Add tags to report: the autocomplete text for tag suggestions is super tiny!
- Reports list > Tags list:
    - "x" button does nothing for me
    - What should "Clear tags" do? That seems dangerous, but it doesn't do anything for me
- Incident details: Tags for reports list are gray squares, with no color or text (details probably didn't get loaded)

harshilshah4251 commented on public/angular/js/controllers/incidents/report_incident_modal.js:
```diff
@@ -106,12 +130,48 @@ angular.module('Aggie')
+    $scope.saveReport = function(report) {
+      Report.save({ id: report._id }, report, function() {
+      }, function() {
+        flash.setAlertNow("Sorry, but that report couldn't be saved for some reason");
```
NP: Is it necessary to say "for some reason" ? Either we could state the reason or remove that part since it doesnt really add anything..up to you.

limmabean replied:
Fixed in: https://github.com/TID-Lab/aggie/commit/556591795b099aa5367e620965e5268e40adac7a

Something I also found strange. This wasn't made by me and is messaging used throughout aggie. I shortened it.

harshilshah4251 commented on public/angular/js/controllers/sources/index.js:
```diff
@@ -10,7 +10,7 @@ angular.module('Aggie')
   function($scope, $rootScope, flash, sources, Source, Tags, Socket) {
     $scope.sources = sources;
-
+    console.log($scope.sources);
```
is this for testing or production?

limmabean replied:
Fixed in https://github.com/TID-Lab/aggie/commit/556591795b099aa5367e620965e5268e40adac7a

It was for testing.

harshilshah4251 commented on public/angular/js/controllers/tags/report_tags_modal.js:
```diff
@@ -0,0 +1,250 @@
+      };
+
+      $scope.onInvalidSMTCTagAdd = function(e, addedTag) {
+      };
```
why empty functions?

limmabean replied:
Fixed in: https://github.com/TID-Lab/aggie/commit/556591795b099aa5367e620965e5268e40adac7a

I thought they might be useful in the future. I'm taking them out because its excess code and I have left a comment with documentation of the tagify library.

harshilshah4251 commented on public/angular/js/controllers/tags/report_tags_modal.js:
```diff
@@ -0,0 +1,250 @@
+        // Pick one of the sources that has a media type. For now, it happens that
+        // if a report has multiple sources, they all have the same type, or are
+        // deleted
+        if ($scope.selectedReport.metadata.platform === "Facebook") {
```
I was thinking we should make some constants.js(or something similar) file that contains all these types of Constant String values. Because I also have been hardcoding strings which I dont think is a good approach. I will try to add that in next PR.

limmabean replied:
Sounds good. Please do.

harshilshah4251 reviewed (commented):
Good work..Is there a limit on how many tags tracker can add? And do tags work across FB, twitter etc?

mb371 commented:
Bravo Matthew! I am excited to try this out and would like to get access to a test instance if that is possible please. As Kevin says, we can play around with it internally before we do the release.
Thanks,
Mike

From: Kevin Cooper <notifications@github.com>
Reply-To: TID-Lab/aggie <reply@reply.github.com>
Date: Friday, August 28, 2020 at 12:56 PM
To: TID-Lab/aggie <aggie@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [TID-Lab/aggie] Tag Page, Group Report Tag Add, Individual Report Tag Add (#361)


@cooperka commented on this pull request.

We can deploy this after rough code review so that folks can use it right away; before we merge to develop though, I do think it needs more tests.

Nice work overall, this has been a long time coming and I'm so excited!

List of random things I found, nothing specifically blocking release:

  *   Add tag > Submit: button doesn't work if you haven't chosen a color (error message would be helpful, and/or default to a random color to make it one less step)
  *   Edit tag > Save:
     *   Old info doesn't update until you manually refresh the page
     *   Alert message says smtcTag.update.success instead of real text
  *   Reports list > Add tags to report: the autocomplete text for tag suggestions is super tiny!
  *   Reports list > Tags list:
     *   "x" button does nothing for me
     *   What should "Clear tags" do? That seems dangerous, but it doesn't do anything for me
  *   Incident details: Tags for reports list are gray squares, with no color or text (details probably didn't get loaded)

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/TID-Lab/aggie/pull/361#pullrequestreview-477885645>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/APJQE6D2NZVWKYY7XK4L5T3SC7OTVANCNFSM4QOJGOLQ>.

Commit:
Fixed Messaging, Removed Debug code and Unnecessary code, and Updated Locale Jsons

I changed messaging from "sorry that report couldn't be saved for some reason" to "sorry that report couldn't be saved." I also added messages for updated tags and removed a bunch of debugging code I used and forgot to remove.

Commit:
Made the smtcTags toggle-controlled

Originally the tags were controlled using an add/remove system. This was clearly confusing to a certain extent. I have changed the system. I also added a bunch of comments to the index.html controller in the reports.

Commit:
Clicking a Tags on the Report Table Removes It

In addition to the toggle setting I have added a way of removing individual tags on each individual report by clicking a tag in the Tag column of the reports table. This allows for more fine tuned editing of tags without opening the tags modal.

limmabean commented:
Working through these issues:

Fixed in (https://github.com/TID-Lab/aggie/commit/556591795b099aa5367e620965e5268e40adac7a):
>   * Alert message says `smtcTag.update.success` instead of real text


Fixed in (https://github.com/TID-Lab/aggie/commit/f93ed3e8ff9a03895bd6fb459715422aa03da7b9)
>   * What should "Clear tags" do? That seems dangerous, but it doesn't do anything for me
>   * "x" button does nothing for me
> * Incident details: Tags for reports list are gray squares, with no color or text (details probably didn't get loaded)

harshilshah4251 reviewed (approved):
LGTM

Commit:
Tags are now editable from the Report Show Page

Removed a lot of console.logs and corrected the messaging for when no tags are made yet.

Commit:
Moved Spectrum Script inside AngularJS Controller

Removed Script tag as it's not good code.

Commit:
Tag validation Clarifications

Made it clear what fields were required to create a tag and also created a color palette that imitates trello.

Commit:
Fix Updates for Tags

Tags now update without refreshing on edit and creation.

Commit:
Responsive table on SMTCtag modal

The table on the modal is too small sometimes and leaks out of the modal. Using a responsive table fixes this problem.

limmabean commented:
Fixed in (https://github.com/TID-Lab/aggie/commit/c365eb46efaf36eec8160a86d8e0ba73dee59e05):
> * Tags page > Add tag > Submit: button doesn't work if you haven't chosen a color (error message would be helpful, and/or default to a random color to make it one less step)

Fixed in (https://github.com/TID-Lab/aggie/commit/17e2af691e0a514e8bf0f58894c8d2cb1111d411):
>   * Old info doesn't update until you manually refresh the page

Not fixed (Not important for PR)
> * Reports list > Add tags to report: the autocomplete text for tag suggestions is super tiny!

Commit:
One last console.log to remove

cooperka commented:
Thanks again for the hard work! I know you're under some time pressure here but here's the result of round 2 testing:

1. New tag error message is ever-present now, but still nothing changes on submit (I would expect to not see the error until I press submit, then the error appears)

    <img width="120" alt="Screen Shot 2020-08-31 at 12 00 08" src="https://user-images.githubusercontent.com/2047062/91756568-c95f5c80-eb81-11ea-94f1-a018977804eb.png">

1. Gray squares for tags on incident view still

    <img width="80" alt="Screen Shot 2020-08-31 at 11 59 17" src="https://user-images.githubusercontent.com/2047062/91756591-d2502e00-eb81-11ea-9cb3-c8f767d2bb73.png">

1. Tag edit refreshes the UI successfully now, but add a new tag > edit that tag **before refreshing the page** > details are blank in the modal

1. Similar, add a new tag > edit that tag **before refreshing the page**  > refresh the page > there are 2 tags instead of updating the old one

1. Custom tag color isn't recognized (create/edit tag and use the advanced color chooser > just appears as gray)

1. Minor, but quick win: Long tag covers up checkbox making it unclickable (could just expand the checkbox click target to the whole table cell, that would make it easier in general)

    <img width="200" alt="Screen Shot 2020-08-31 at 12 21 38" src="https://user-images.githubusercontent.com/2047062/91758269-923e7a80-eb84-11ea-93cd-cc81c278a30f.png">

1. Tag "x" button is still visible in batch mode (new toggle mode works for the regular view)

Commit:
Form Validation Errors Show when Violated. Colorpicker starts with no color or edited tag color.

This fixes the validation errors from showing all the time. Color Picker now starts with color or edited tag color.

Commit:
Move to Laptop

Commit:
Merge branch 'ML-tags-Actual' of https://github.com/TID-Lab/aggie into ML-tags-Actual

Commit:
Fixed incidents page, bugs surrounding editing tags, custom colors not working

I think this is the last commit necessary before PR is merged.

Commit:
Merge branch 'develop' into ML-tags-Actual

# Conflicts:
#	config/crowdtangle_list.json

Commit:
Merge branch 'develop' into ML-tags-Actual

cooperka committed:
Merge branch 'develop' into ML-tags-Actual

Commit:
Change to Using Aggie Docs to include SMTCTags

Commit:
Merge remote-tracking branch 'origin/ML-tags-Actual' into ML-tags-Actual

Commit:
Update Batch Page Tag Sidebar to Match Report Page

Commit:
Update Toggle Tag on Batch to Match Report Page

cooperka committed:
Optimize new images

cooperka reviewed (approved):
Amaaaaazing 🚀 and thank you for the docs too. I tested locally and all issues are fixed.

I don't want to block this any further, I think we can squash and merge. That said, for posterity, what do you think you'd want to change if you **hypothetically** had an extra week to work on this, e.g. any specific spec coverage? Let's document what's needed while it's fresh in your head because we may not get around to it soon.
